### PR TITLE
fix(CheckTree): searchKeyword is not updated as expected

### LINKED ISSUE #3196: CheckTree: props "searchKeyword" and "searchBy" filter the tree using the previous react state value
### What version of rsuite are you using?

^5.31.0

### What version of React are you using?

^18.2.0

### What version of TypeScript are you using (if any)?

_No response_

### What browser are you using?

Chrome

### Describe the Bug

The internal search performed on CheckTree when using either the `searchKeyword` or the `searchBy` prop is "delayed": it is possible that the component is using the previous state to filter the tree instead of the current text value.

Example:

- `searchKeyword` = "a" -> tree values are filtered by ""
- `searchKeyword` = "ab"  -> tree values are filtered by "a"
- `searchKeyword` = "abc"  -> tree values are filtered by "ab"
- `searchKeyword` = "ab"  -> tree values are filtered by "abc"

### Expected Behavior

The CheckTree should be filtered using the current (updated) value passed using the `searchKeyword` prop.

#### In CodeSandbox:
```
import React, { useState } from "react";
import ReactDOM from "react-dom";
import "./styles.css";
import { CheckTree } from "rsuite";
import { mockTreeData } from "./mock";

const data = mockTreeData({
  limits: [3, 3, 4],
  labels: (layer, value, faker) => {
    const methodName = ["jobArea", "jobType", "firstName"];
    return faker.name[methodName[layer]]();
  }
});

const App = () => {
  const [textFilterValue, setTextFilterValue] = useState("");

  function onChange(e) {
    setTextFilterValue(e.target.value);
  }

  return (
    <>
      <input value={textFilterValue} onChange={onChange} />
      <CheckTree data={data} defaultExpandAll searchKeyword={textFilterValue} />
    </>
  );
};

ReactDOM.render(<App />, document.getElementById("root"));
```

### PR DESCRIPTION
**Pull request by superman66**
## Issue
The searchKeyword prop of the CheckTree component is not updating as expected. 

close #3196 

## Cause
The problem is rooted in the logic responsible for updating the searchKeyword within the CheckTree component.

## Solution
In this pull request, I address the issue by reviewing and adjusting the code related to the searchKeyword functionality in the CheckTree component. This ensures that the searchKeyword state is updated appropriately and reflects user input accurately.

## Changes Made
I reviewed the relevant code sections responsible for handling the searchKeyword and made the necessary modifications to enable its proper updating. This included checking for event listeners, data binding, and any conditional statements that might be affecting the searchKeyword state.

## Testing
I thoroughly tested the changes by simulating various search scenarios and confirming that the searchKeyword state now updates correctly. And I also add a test case for this problem.

Screenshots 

https://github.com/rsuite/rsuite/assets/12592949/44448f45-5f56-4b98-aeef-1542d8d270e8

### CODE REVIEW

File: src/utils/treeUtils.ts [hunk 1]
```diff
@@ -930,25 +930,30 @@ export function useTreeSearch<T>(props: TreeSearchProps<T>) {
...
+    setFilteredData(filteredData);
+    setSearchKeyword(searchKeyword);
+    event && callback?.(searchKeyword, filteredData, event);
+  };
+
+  useEffect(() => {
+    handleSearch(searchKeyword ?? '');
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [searchKeyword]);
+
```
Comment (SevenOutman) [inline src/utils/treeUtils.ts:949]: Why not derive `filteredData` from `data` and `searchKeywordState`?

Comment (superman66) [reply src/utils/treeUtils.ts:949]: You are right. `filtereredData` should be derived from `data` and `searchKeywordState`.

As the useTreeSearch hook is utilized by both the CheckTree and Tree components, I intend to create a new Pull Request with the goal of optimizing it.

**Review by SevenOutman — APPROVED** [review]

**Comment by zeari** [comment]
Hey, thanks for fixing this! 
Could you tell me how long until this fix is released?

**Comment by superman66** [comment]
@SevenOutman Will we release a version this week?

**Comment by SevenOutman** [comment]
Yes it's going to be released this Friday.

**Pull request merged**